### ISSUE TITLE
Schema cccv1p0.xsd is not a valid XSD File

### ISSUE THREAD

**Issue by CreepyGnome** [instructure/canvas-lms#1134]
Your Common Cartridge Resource associatedcontent/imscc_xmlv1p1/learning-application-resource Schema is invalid.

The file is located in your source here: 

canvas-lms/lib/cc/xsd/cccv1p0.xsd

Also hosted here:

http://canvas.instructure.com/xsd/cccv1p0.xsd

I opened this in XML Spy and it threw a fit on the file as it isn't a valid XSD file.  So I found an online validator to check against that you can use to identify and fix your XSD file.

If you go here:

http://www.utilities-online.info/xsdvalidation/

Then past your contents of the above XSD file into right side labeled XSD Schema and then click the check XSD Validity you can get a list of all the errors in the file, I would copy and paste them but the site doesn't allow it for some reason.

There are at least 7 it can find on its first pass.

- 3 Elements defined a type and defined a complex type which is not allowed. I removed those
- 1 Elements defined a type and defined a simple type which is not allowed. I removed those
- 1 Attributes defined a type and defined a simple type which is not allowed. I removed those
- 1 Sequence under a complexType
- 1 all under a complexType

The first 5 issues are easy just remove the type as they are defining them as xs:string while also defining a complex or simple type. If it was just those I would have made a PR for them, however since the other two require some remove as it will impact what you think your XSD is doing someone at Instructure should review and fix all them.

Without a valid XSD file it cannot be used to ensure the extra files conform to it.  I hope someone can fix this quickly. Thanks.

**Comment by omarkhan** [comment]
Thanks for this. We're aware of the issue, and we're working on a fix. We should have it merged soon.

**Comment by CreepyGnome** [comment]
Any status on this? I haven't seen any changes to the file yet in master.

**Comment by omarkhan** [comment]
We are working on a fix. It hasn't been merged yet, there are still some issues with it.

**Comment by simonista** [comment]
Looks like this was fixed in eba07fd3fe5c53eb0b08d3fadc4886d36a77b48e.  Thanks again for the report!

**Comment by CreepyGnome** [comment]
@simonista you guys close issues before they are released?  If monitoring the issue isn't how I can be notified when a fix is released to stable and/or production environment how can else can I find out?

That XSD file in question has not yet been made public via its public URL you host  yet:

http://canvas.instructure.com/xsd/cccv1p0.xsd

This URL is what is used as a reference in various XSD files as well so it is a production URL.

The only place to get the current changed file is on the master branch.   If I created a file and referenced this schema location that is public and someone tried to use that file it would fail still today until.

**Comment by simonista** [comment]
Yeah, we close issues when they are merged to the master branch.  This is how most other projects I've interacted with work (probably at least partially because this is how Github encourages you to work, see https://help.github.com/articles/closing-issues-using-keywords/, although we're different since we're using Gerrit and Jira internally and then mirroring code to Github.)

You can watch the beta and stable branches to see when a commit makes it to our beta and production accounts respectively, or you can watch the release notes (see https://community.canvaslms.com/docs/DOC-1271-all-about-canvas-release-notes).  There's also a deploy schedule posted there: https://community.canvaslms.com/docs/DOC-1384-what-is-the-canvas-release-schedule-for-beta-production-and-test-environments, but the tl;dr is that we deploy master -> beta and beta -> production every 3 weeks, so you can expect code merged to master to be in production in 0-6 weeks.

Lastly, a meta-comment.  I feel like the tone of your comments here has been disappointing. You've raised some legitimate issues and I totally understand feeling frustrated by them, but keep in mind that open source support is provided by volunteers. Making people feel unhappy only makes us less interested in digging into the issues you raise.

**Comment by CreepyGnome** [comment]
@simonista wow so much to unpack from that last comment.

Many projects I've interacted with on GitHub use a CICD pipeline which is encouraged by GitHub which means once it's merged to master it gets released very quickly. This would be when I would think a public issue reported by an external source is encouraged to be closed since typically in most mature GitHub projects merging to main also means it was released due to their CICD pipeline.

Also, since you are bringing it up you do not really follow GitHub best practices.

I do watch the Stable branch and I get release note notifications to know what has changed and been pushed to stable.

Just because you have an open source project doesn't mean you get to claim you are supported by volunteers. You make it sound like a majority of your contributes are from unpaid people who received no financial compensation for supporting Canvas. I would bet almost all your contributions in the code base are from Instructure employees or a partner who has a financial reason to contribute. 

As far as my tone, I guess it is that I am disappointed. Instructure is a large company with a lot of developers who are not volunteers, like yourself. As far as I have been able to tell all the interactions with Instructure's GitHub repo have been with their Employees.  Keep in mind when I get short, unhelpful, or hostile responses from someone its hard not to be disappointed or have a tone that is leaning in a negative direction when I respond to them.

I am not sure how I am making people feel unhappy? Is it that I am finding bugs in Canvas? Is it that I find bugs in the documentation? Is it that I am monitoring the bugs? Is it that I care about Canvas and want to help it improve by making it easier for people outside of Instructure to consume? Is it that I am trying to actually use the opensource Canvas LMS as an actual LMS and not the cloud version?

I am serious as to wanting to know how am I making employees of Instructure who work on Canvas unhappy I would like to know?

If you don't want to know about bugs or want any feedback into helping to improve the open source experience of Canvas let me know I will stop reporting issues. I felt my experience would be useful to you and was volunteering it.

**Comment by bootstraponline** [comment]
Hi,

It's common on GitHub to close issues after they have been fixed on master. I understand this makes it difficult to know when fixes are available in a release.

I appreciate that you've been finding bugs in Canvas. Supporting issues on GitHub is done on a volunteer basis. I am having trouble encouraging our engineers to participate on the issue tracker because it's not a welcoming environment. If we're not able to change that for the better, then that support will go away.

I have been trying to increase our investment in open source. I'd love if everyone could help us ensure that GitHub issues remains a productive and pleasant use of our time.

**Comment by omarkhan** [comment]
Perhaps I can clarify. Instructure chooses to release source code for Canvas and other projects under an open source license because it allows us to improve the quality of our software through contributions from community members such as yourself. Many contributors, yourself included, have helped us immensely by providing bug reports and often fixes as well. Most of us here are Instructure employees who engage with the community on a voluntary basis - this is work we do in addition to our day jobs. We do this because we believe in open source and want to promote a more open culture within the company, but it's a hard case to make when contributors are unfriendly or behave with a sense of entitlement. So be nice!

**Comment by CreepyGnome** [comment]
Thanks, @bootstraponline @omarkhan for the information I do appreciate it.

**Comment by bootstraponline** [comment]
I recommend this book:

- `Crucial Conversations Tools for Talking When Stakes Are High, Second Edition`
 https://www.amazon.com/Crucial-Conversations-Talking-Stakes-Second-ebook/dp/B005K0AYH4/ref=mt_kindle?_encoding=UTF8&me=

It's been helpful to me in improving perceptions when talking with people.

**Comment by CreepyGnome** [comment]
Thanks @bootstraponline I am already actually reading that book. 

However, that doesn't help find where I have offended someone in an issue in this repo with a "not nice" comment that wasn't in response to a "not nice" comment directed at me?  I am serious to know where this occurred so I can see what words I used incorrectly.  I learn best with examples, and when I made a mistake I learn by reviewing the mistake so I know what not to do or how to correct going forward.

As far as I can see from review everything in this and other issues in this repo from me seem fine and there is no noticable positive or negative slant. The only exception may be the one response to Simon which was in response to a lot if issues that I tried to make a mater of fact response to address issues Simon brought up that where not accurate or seemed negatively directed at me.
